### ISSUE TITLE
Online DPO trains every tie, judge failure and swap-order disagreement as "the second completion wins": with the judge down, a whole run learns arbitrary labels

### ISSUE THREAD

**Issue by MakazhanAlpamys** [MakazhanAlpamys/Soup#1225]
Found in a code review; reproduced on `main` at `99d553f5` with trl 0.29.1 (CPU, `hf-internal-testing/tiny-random-gpt2`).

## What happens

Soup's pairwise judge returns `-1` whenever it cannot name a winner: the judge call failed, the reply did not parse, or the two orders of the swap check disagreed.

```python
import httpx

from soup_cli.eval.judge import JudgeEvaluator, make_soup_pairwise_judge

evaluator = JudgeEvaluator(provider="ollama", model="llama3.1")


def server_down(prompt):
    raise httpx.ConnectError("connection refused")


evaluator._call_llm = server_down  # the judge server is not running
print(make_soup_pairwise_judge(evaluator).judge(["p"] * 3, [["first", "second"]] * 3))


class AlwaysA:  # a reachable judge with a position bias
    def compare_pair(self, prompt, a, b):
        return 0


print(make_soup_pairwise_judge(AlwaysA()).judge(["p"] * 3, [["first", "second"]] * 3))
```

```
[-1, -1, -1]
[-1, -1, -1]
```

What trl 0.29.1 does with a `-1` was measured on the real trainer: `OnlineDPOTrainerWrapper.setup()` builds trl's `OnlineDPOTrainer` with Soup's judge (`judge=` path, `_trl_has_judges()` is `True` on 0.29.1). Then one real `training_step` is run on the same batch, same seed and same weights (so generation is identical), with three judges: Soup's judge with the server down, a judge that returns 1 ("second wins") for every pair, and one that returns 0 ("first wins").

```
ranks the Soup judge returned during its step: [-1, -1, -1, -1]
arm                              loss       |grad|  logps/chosen  logps/rejected
soup judge, server down      0.173287     0.046399      -41.3589        -41.3277
rank 1 (second wins)         0.173287     0.046399      -41.3589        -41.3277
rank 0 (first wins)          0.173287     0.046399      -41.3277        -41.3589
grad(server down) == grad(rank 1): True
grad(server down) == grad(rank 0): False
cosine(grad(rank 1), grad(rank 0)) = -1.0
```

A dropped pair would contribute a zero gradient. Instead, the step with an unreachable judge is **bit-identical** to a step with a judge that declares the second completion the winner of every pair, and exactly opposite (cosine -1.0) to "first wins".

<details><summary>Script for the training-step comparison</summary>

```python
import torch
import trl

import soup_cli.trainer.online_dpo as od
from soup_cli.config.loader import load_config_from_string

# `evaluator` is the server-down JudgeEvaluator from the snippet above


class Const(trl.experimental.judges.BasePairwiseJudge):
    def __init__(self, rank):
        self.rank = rank

    def judge(self, prompts, completions, shuffle_order=True):
        return [self.rank] * len(prompts)


cfg = load_config_from_string(
    "base: hf-internal-testing/tiny-random-gpt2\ntask: online_dpo\n"
    "data:\n  train: x.jsonl\n  max_length: 64\n"
    'training:\n  online_dpo_judge: "ollama://llama3.1"\n  epochs: 1\n  batch_size: 4\n'
    "  online_dpo_max_new_tokens: 6\n  lr: 1e-4\n  quantization: none\noutput: ./out\n"
)
dataset = {"train": [{"messages": [{"role": "user", "content": t}]}
                     for t in ("hi there friend", "hello", "what is up", "tell me a story")]}
od._ONLINE_DPO_JUDGE_OVERRIDE = evaluator  # the test seam the wrapper already has
wrapper = od.OnlineDPOTrainerWrapper(cfg, device="cpu")
wrapper.setup(dataset)
od._ONLINE_DPO_JUDGE_OVERRIDE = None
trainer, model = wrapper.trainer, wrapper.trainer.model
soup_judge = trainer.judge
batch = next(iter(trainer.get_train_dataloader()))
trainable = [p for p in model.parameters() if p.requires_grad]


def one_step(judge):  # training_step computes gradients only; weights do not change
    model.zero_grad(set_to_none=True)
    trainer.judge = judge
    torch.manual_seed(1234)
    loss = trainer.training_step(model, batch)
    return loss, torch.cat([p.grad.flatten() for p in trainable if p.grad is not None])


_, g_down = one_step(soup_judge)
_, g_one = one_step(Const(1))
_, g_zero = one_step(Const(0))
print(torch.equal(g_down, g_one), torch.equal(g_down, g_zero),
      torch.nn.functional.cosine_similarity(g_one, g_zero, dim=0).item())
```

</details>

## Why

- `src/soup_cli/eval/judge.py:306-316` (`compare_pair`) turns any exception into `-1` and logs it at DEBUG only. `pairwise_compare` (`judge.py:389-421`) returns `-1` for a tie, a failure, or a disagreement between the two orders.
- The docstring of `make_soup_pairwise_judge` (`judge.py:533-534`) says "TRL treats -1 as a dropped sample". On trl 0.29.1 that is false. `OnlineDPOTrainer.training_step` (`trl/experimental/online_dpo/online_dpo_trainer.py:1363-1374`) builds `mask = rank == 0`, so `-1` becomes "second is chosen, first is rejected", and `losses.mean()` averages over every pair.
- trl's own contract says the caller must handle `-1` (`BasePairwiseJudge.judge` docstring, `trl/experimental/judges/judges.py:151-155`: "the caller should handle these invalid indices appropriately"), and `OnlineDPOTrainer` does not.
- Soup takes this path on every supported trl: `pyproject.toml:112` pins `trl>=0.29.0,<1.0.0`, and the wrapper picks `judge=` whenever trl accepts it (`src/soup_cli/trainer/online_dpo.py:120-127` and `:424-431`).

With an unreachable judge, a 401, or replies that do not parse, every pair of every batch is `-1`. The run then optimises toward whichever of two i.i.d. samples happened to be generated second. The step above raises nothing, and the failure is logged only at DEBUG, so nothing stops the run. With a reachable but position-biased judge, exactly the pairs the swap check refuses to trust still train, all in one direction.

## Fix path

Drop the pairs the judge could not rank. Override `training_step` in a thin `OnlineDPOTrainer` subclass (or wrap the per-pair loss) so pairs with rank `-1` get zero weight, the mean is taken over the kept pairs, and a step with no kept pair contributes nothing; the normalisation under gradient accumulation has to use the kept count. Mapping `-1` to a coin flip or to `0` is *not* a fix: it still trains on noise. Also make failures visible: log the per-step invalid rate as a metric, warn at WARNING level, and abort with a message naming the judge URL when the judge fails for several consecutive batches (for example, every pair `-1` for N steps). Correct the docstring.

The trl 1.x `reward_funcs` path, outside today's trl bound, has the mirror-image problem when read from the code (not executed here). `make_judge_reward_func` scores a failed call as 0.0 (`judge.py:506-508`), and trl ranks with `first_half >= second_half`, so a failure there means "first wins".

## Acceptance

- [ ] The step comparison above becomes a CPU test: with ranks `[-1] * B`, the step contributes a zero gradient (or is skipped), and it no longer equals the rank-1 gradient.
- [ ] Mixed ranks `[0, -1, 1, -1]`: the gradient equals that of a batch containing only the two ranked pairs.
- [ ] A server-down judge for a whole (tiny) run ends with an error naming the judge URL, or at minimum a WARNING with the invalid rate, instead of a normal completion.
- [ ] A position-biased judge (always `A`) produces dropped pairs and a logged invalid rate.
- [ ] The `make_soup_pairwise_judge` docstring states what actually happens.
